### PR TITLE
add login/signup validation

### PR DESCRIPTION
**Pull request by armanmoztar**
Signup:
- [ ] Check if a user already exists in db

Login:
- [ ] Check if password is incorrect

### CODE REVIEW

File: backend/src/controllers/userController.js [hunk 1]
```diff
@@ -42,6 +42,22 @@ export default class UserController {
...
+          reject({ error: err });
+        }
+        if (res && res.length > 0 && res[0].email === req.body.email) {
+          reject({ error: "Email already exists" });
+        }
+        resolve(res);
+      });
+    });
+  }
+
```
Comment (armanmoztar) [inline backend/src/controllers/userController.js:60]: Wasn't really sure how to connect endpoint to frontend part of SignUpForm

File: frontend/src/App.js [hunk 2]
```diff
@@ -65,10 +66,11 @@ function App() {
           <Route path="/pdf" element={(currentUser && currentUser.employee) ? <InvoicePreview /> : (currentUser ? <Setting /> : <LoginForm from={window.location.pathname} />)} />
           <Route path="/login" element={<LoginForm />} />
           <Route path="/signup" element={<SignUpForm />} />
+          {/* <Route path="/forgotpassword" element={<ForgotPasswordForm />} /> */}
```
Comment (Harin329) [inline frontend/src/App.js:69]: We can uncomment all these and add it in, if there's no entry point, it won't be accessible anyways so no need to comment

File: backend/src/controllers/userController.js (hunk 1 above)
Comment (Harin329) [reply backend/src/controllers/userController.js:60]: What I would do is make an endpoint that is a get, which takes in a param as a user email. And then you'll basically just run a SELECT * from users where user_email = <EMAIL>; and basically if the result is not 0, the endpoint can return a false and that will ultimately fail the signup.

Or, you can do the check directly in the POST user endpoint, and in the controller for that, do a check to see if the user exists, and if so, fail that POST signup user endpoint.

Comment (armanmoztar) [reply backend/src/controllers/userController.js:63]: If I go with the first option, do I need to modify the usersProc.sql file? I also noticed there's loadSingleUser which has SELECT * FROM users WHERE email = _email;

Comment (Harin329) [reply backend/src/controllers/userController.js]: oh yeah then I think you can just reuse that one, and not have to deal with the database, you can use that model call as a part of the signup process and do the check there

File: frontend/src/assets/LoginBackground.png
Comment (Harin329) [inline frontend/src/assets/LoginBackground.png:1]: Also I don't think we want to be changing this image

File: backend/src/controllers/userController.js [hunk 3]
```diff
@@ -61,6 +61,19 @@ export default class UserController {
     });
   }
 
+// check if email already exists in database
+  checkEmail(email) {
+    return new Promise((resolve, reject) => {
+      const UserModel = new User();
+      UserModel.checkEmail(email, (err, result) => {
```
Comment (Harin329) [inline backend/src/controllers/userController.js:68]: essentially this UserModel.checkEmail part can be called in the signup, and we will check the returned result to see if it's empty or not and decide if we proceed

File: backend/src/routes/userRoute.js [hunk 4]
```diff
@@ -51,6 +51,18 @@ router.get("/", authorize(), (_, res) => {
     });
 });
 
+router.get("/checkEmail", authorize(), (req, res) => {
```
Comment (Harin329) [inline backend/src/routes/userRoute.js:54]: this route is therefore unnecessary if we do it as a part of the user signup

**Review by Harin329 — CHANGES_REQUESTED** [review]

File: backend/src/controllers/userController.js (hunk 3 above)
Comment (armanmoztar) [reply backend/src/controllers/userController.js:68]: Do we import User from the backend if I add the UserModel.checkEmail in signup
Screenshot 2023-04-10 at 2 35 57 PM

Comment (Harin329) [reply backend/src/controllers/userController.js:68]: oh no you would check it all on the backend and you can return with a different 400 error message and the frontend checks for that. The frontend should never import from the backend like that

Comment (armanmoztar) [reply backend/src/controllers/userController.js:68]: I think I almost figured out a way but when I try to send an email query on Postman I get "message": "Unauthorized Error". 
Screenshot 2023-04-14 at 6 30 49 PM
Screenshot 2023-04-14 at 6 40 18 PM

Comment (Harin329) [reply backend/src/controllers/userController.js:68]: It's possible that the fetch user call your making requires you to be auth'd, I would check that first.

**Comment by Harin329** [comment]
Completed via https://github.com/ubclaunchpad/labby/commit/373bcb9446581b341d5fa7df9f9313df0bf680d2 and https://github.com/ubclaunchpad/labby/commit/bb22f1e5f6c7a85be00722da44949f38872d30d6

**Pull request closed without merging**